### PR TITLE
Grid and Form Permissions

### PR DESCRIPTION
**Pull request by samargupta007**

### CODE REVIEW

File: src/lib/components/utils.js [hunk 1]
```diff
@@ -11,5 +11,20 @@ const utils = {
         OverrideDateFormat: "DD-MMM-YYYY"
     }
 }
+export const getPermissions = (userData, model, fallbackPermissions = {}) => {
```
Comment (Pramodspraxa) [inline src/lib/components/utils.js:14]: Code should be like below - 

```
export const getPermissions = (userData, model, fallbackPermissions = {}) => {
    const userPermissions = userData?.filter(item => item.Module === model.module) || [];
    const permissionsToUse = userPermissions.length ? userPermissions[0] : {};
    if(!userPermissions.length) {
	return { canAdd: fallbackPermissions.add || false, canEdit: fallbackPermissions.edit || false, canDelete: fallbackPermissions.delete || false };
    }
    return {
        canAdd: permissionsToUse.Permission2 === 1,
        canEdit: permissionsToUse.Permission3 === 1,
        canDelete: permissionsToUse.Permission4 === 1 
    };
};
```

File: src/lib/components/Form/Form.js [hunk 2]
```diff
@@ -43,6 +44,9 @@ const Form = ({
     const fieldConfigs = model?.applyFieldConfig ? model?.applyFieldConfig({ data, lookups }) : defaultFieldConfigs;
     let gridApi = `${url}${model.api || api}`
     const { mode } = stateData.dataForm;
+    const userData = stateData.getUserData;
+    const fallbackPermissions = {edit: permissions.edit,delete: permissions.delete};
```
Comment (Pramodspraxa) [inline src/lib/components/Form/Form.js:48]: add permission is misssing. 
If someone start to use directly then it by default will come as false always.

File: src/lib/components/Grid/index.js [hunk 3]
```diff
@@ -212,6 +213,9 @@ const GridBase = memo(({
     const currentPreference = stateData?.currentPreference;
     const tablePreferenceEnums = stateData?.gridSettings?.permissions?.tablePreferenceEnums;
     const emptyIsAnyOfOperatorFilters = ["isEmpty", "isNotEmpty", "isAnyOf"];
+    const userData = stateData.getUserData;
+    const fallbackPermissions = {add: effectivePermissions.add,edit: effectivePermissions.edit,delete: effectivePermissions.delete};
```
Comment (Pramodspraxa) [inline src/lib/components/Grid/index.js:217]: formatting missing.

Comment (Pramodspraxa) [reply src/lib/components/Grid/index.js:217]: fallbackPermissions
name should be userDefinePermissions or something.

File: src/lib/components/Grid/index.js [hunk 4]
```diff
@@ -671,8 +675,8 @@ const GridBase = memo(({
             >
                 {model.gridSubTitle && <Typography variant="h6" component="h3" textAlign="center" sx={{ ml: 1 }}> {(model.gridSubTitle)}</Typography>}
                 {currentPreference && <Typography className="preference-name-text" variant="h6" component="h6" textAlign="center" sx={{ ml: 1 }} >Applied Preference - {currentPreference}</Typography>}
-                {(isReadOnly || (!effectivePermissions.add && !forAssignment)) && <Typography variant="h6" component="h3" textAlign="center" sx={{ ml: 1 }} > {isReadOnly ? "" : model.title}</Typography>}
-                {!forAssignment && effectivePermissions.add && !isReadOnly && !showAddIcon && <Button startIcon={!showAddIcon ? null : <AddIcon />} onClick={onAdd} size="medium" variant="contained" className={classes.buttons} >{model?.customAddTextTitle ? model.customAddTextTitle : ` ${!showAddIcon ? "" : `${"Add"}`} ${model.title ? model.title : 'Add'}`}</Button>}
+                {(isReadOnly || (!canAdd && !forAssignment)) && <Typography variant="h6" component="h3" textAlign="center" sx={{ ml: 1 }} > {isReadOnly ? "" : canAdd ? model.title : ''}</Typography>}
```
Comment (Pramodspraxa) [inline src/lib/components/Grid/index.js:678]: this extra condition should never occur - 
: canAdd ? model.title : ''

**Review by Pramodspraxa — CHANGES_REQUESTED** [review]

**Review by manojpandeyDev18 — APPROVED** [review]

**Pull request merged**